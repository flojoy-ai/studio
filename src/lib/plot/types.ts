export interface Drawable {
  draw: () => void;
}
