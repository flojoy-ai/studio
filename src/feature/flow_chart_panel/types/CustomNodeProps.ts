export type ElementsData = {
  id: string;
  label: string;
  func: string;
  type: string;
  running?: boolean;
  failed?: boolean;
  ctrls: {
    [key: string]: {
      functionName: string;
      param: string;
      value: string | boolean | number;
    };
  };
  inputs?: Array<{ name: string; id: string; type: string }>;
  selects?: any;
  selected?: boolean;
  handleRemove?: (nodeId: string) => void;
};

export interface CustomNodeProps {
  data: ElementsData;
}
