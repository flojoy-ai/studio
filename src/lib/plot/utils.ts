export function clamp(x: number, lo: number, hi: number) {
  return Math.min(Math.max(x, lo), hi);
}
