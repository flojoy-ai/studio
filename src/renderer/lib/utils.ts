import { type ClassValue, clsx } from "clsx";
import { twMerge } from "tailwind-merge";
import { z } from "zod";
import * as E from "fp-ts/Either";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function tryParse<Z extends z.ZodTypeAny>(z: Z) {
  return (val: unknown) => {
    const res = z.safeParse(val);
    if (res.success) {
      return E.right(res.data as z.infer<Z>);
    }
    return E.left(res.error);
  };
}
