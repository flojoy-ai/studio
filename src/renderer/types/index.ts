export type {
  ALL_DC_TYPE,
  DataContainer,
  DataContainerType,
} from "./dataContainer";
export type { CustomNodeProps, BlockData } from "./node";
export type {
  DataContainer2PlotlyProps,
  OverridePlotData,
  PlotProps,
} from "./plotly";
