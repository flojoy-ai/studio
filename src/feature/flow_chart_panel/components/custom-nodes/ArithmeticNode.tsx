import HandleComponent from "@feature/flow_chart_panel/components/HandleComponent";
import { CustomNodeProps } from "@feature/flow_chart_panel/types/CustomNodeProps";
import { useFlowChartState } from "@hooks/useFlowChartState";
import { Box, clsx, createStyles } from "@mantine/core";
import { memo } from "react";
import {
  AddBGTemplate,
  AddSvg,
  MultiplySvg,
  SubSvg,
} from "../../svgs/add-multiply-svg";
import { useNodeStyles } from "../DefaultNode";
import NodeWrapper from "../NodeWrapper";

const useStyles = createStyles((theme) => {
  return {
    arithmeticNode: {
      color:
        theme.colorScheme === "light"
          ? theme.colors.accent1[0]
          : theme.colors.accent2[0],
      background: "transparent",
    },
    operatorIcon: {
      position: "absolute",
      left: 29,
      height: 20,
      width: 20,
    },
  };
});

const ArithmeticNode = ({ data }: CustomNodeProps) => {
  const nodeClasses = useNodeStyles().classes;
  const { classes } = useStyles();
  const { runningNode, failedNode } = useFlowChartState();
  const params = data.inputs || [];

  let operatorIcon: JSX.Element;
  switch (data.func) {
    case "MULTIPLY":
      operatorIcon = <MultiplySvg className={classes.operatorIcon} />;
      break;
    case "ADD":
      operatorIcon = <AddSvg className={classes.operatorIcon} />;
      break;
    case "SUBTRACT":
      operatorIcon = <SubSvg className={classes.operatorIcon} />;
      break;
    default:
      operatorIcon = <Box />;
  }

  return (
    <NodeWrapper data={data}>
      <Box
        className={clsx(
          runningNode === data.id || data.selected
            ? nodeClasses.arithmeticShadow
            : "",
          failedNode === data.id ? nodeClasses.failShadow : ""
        )}
      >
        <Box
          className={clsx(nodeClasses.nodeContainer, classes.arithmeticNode)}
          sx={{
            ...(params.length > 0 && { padding: "0px 0px 8px 0px" }),
          }}
        >
          <AddBGTemplate />
          {operatorIcon}
          <Box
            display="flex"
            h={params.length > 0 ? (params.length + 1) * 40 : "fit-content"}
            sx={{
              flexDirection: "column",
            }}
          >
            <HandleComponent data={data} inputs={params} />
          </Box>
        </Box>
      </Box>
    </NodeWrapper>
  );
};

export default memo(ArithmeticNode);
