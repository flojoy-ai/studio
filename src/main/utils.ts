import * as http from "http";
import * as fs from "fs";
import { isIP } from "net";
import { execCommand } from "./executor";
import { Command } from "./command";
import { app, dialog } from "electron";
import { join } from "path";
import { killCaptain } from "./python";
import log from "electron-log";
import { ChildProcess } from "node:child_process";
import { Err, Result, Ok } from "../types/result";

export const isPortFree = (port: number) =>
  new Promise((resolve) => {
    const server = http
      .createServer()
      .listen(port, "127.0.0.1", () => {
        server.close();
        resolve(true);
      })
      .on("error", () => {
        resolve(false);
      });
  });

export const killProcess = async (port: number) => {
  await execCommand(
    new Command({
      darwin: `kill -9 $(lsof -t -i :${port})`,
      linux: `kill -9 $(lsof -t -i :${port})`,
      win32: `FOR /F "tokens=5" %i IN ('netstat -aon ^| find "${port}"') DO Taskkill /F /PID %i`,
    }),
  );
};

export const ping = async (addr: string) => {
  if (isIP(addr) === 0) {
    throw new Error(`Invalid IP address: ${addr}`);
  }

  return await execCommand(
    new Command({
      win32: `ping -n 1 ${addr}`,
      darwin: `ping -c 1 ${addr}`,
      linux: `ping -c 1 ${addr}`,
    }),
    { quiet: true },
  );
};

export const netstat = async () => {
  return await execCommand(
    new Command({
      win32: "netstat",
      darwin: "netstat",
      linux: "netstat",
    }),
    { quiet: true },
  );
};

export const ifconfig = async () => {
  return await execCommand(
    new Command({
      win32: "ipconfig /all",
      darwin: "ifconfig",
      linux: "ifconfig",
    }),
    { quiet: true },
  );
};

export const writeFileSync = (_, filePath: string, text: string): void => {
  fs.writeFileSync(filePath, text);
};

export const pickDirectory = async (): Promise<string> => {
  const handler = await dialog.showOpenDialog({
    properties: ["openDirectory"],
  });
  return handler.canceled ? "" : handler.filePaths[0];
};

export const getCustomBlocksDir = async () => {
  const filePath = join(app.getPath("home"), ".flojoy/custom_blocks_path.txt");

  if (!fs.existsSync(filePath)) {
    return undefined;
  }

  const blocksPath = fs
    .readFileSync(filePath, { encoding: "utf-8" })
    .toString();

  if (!fs.existsSync(blocksPath)) {
    return undefined;
  }

  return blocksPath;
};

export const cacheCustomBlocksDir = (_, dirPath: string) => {
  const flojoyDir = join(app.getPath("home"), ".flojoy");
  if (!fs.existsSync(flojoyDir)) {
    fs.mkdirSync(flojoyDir);
  }
  const cacheFilePath = join(flojoyDir, "custom_blocks_path.txt");
  fs.writeFileSync(cacheFilePath, dirPath);
};

export const openTestPicker = (): Promise<
  { filePath: string; fileContent: string } | undefined
> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: ["json", "py"],
            name: "Test",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const openFilePicker = (
  _,
  allowedExtensions: string[] = ["json"],
): Promise<{ filePath: string; fileContent: string } | undefined> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: allowedExtensions,
            name: "allowed extensions",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
        resolve(undefined);
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const cleanup = async () => {
  const captainProcess = global.captainProcess as ChildProcess;
  log.info(
    "Cleanup function invoked, is captain running? ",
    !(captainProcess?.killed ?? true),
  );
  if (captainProcess && captainProcess.exitCode === null) {
    const success = killCaptain();
    if (success) {
      global.captainProcess = null;
      log.info("Successfully terminated captain :)");
    } else {
      log.error("Something went wrong when terminating captain!");
    }
  }
};

export const loadFileFromFullPath = async (
  filePath: string,
): Promise<string> => {
  return fs.readFileSync(filePath, { encoding: "utf-8" }).toString();
};

export const saveFileToFullPath = async (
  filePath: string,
  content: string,
): Promise<Result<void>> => {
  try {
    fs.writeFileSync(filePath, content);
    return Ok(undefined);
  } catch (error) {
    return Err(error as Error);
  }
};
