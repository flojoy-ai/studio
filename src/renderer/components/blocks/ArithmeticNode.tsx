import { CustomNodeProps } from "@/renderer/types/node";
import { memo } from "react";
import DefaultNode from "./DefaultNode";

const operatorMap = {
  MULTIPLY: "ร",
  ADD: "+",
  SUBTRACT: "-",
  DIVIDE: "รท",
  ABS: "|x|",
  POWER: "^",
  LOG: "log",
  FLOOR_DIVIDE: "//",
  REMAINDER: "%",
};

const ArithmeticNode = (props: CustomNodeProps) => {
  return (
    <DefaultNode width={72} height={72} {...props}>
      <h2 className="m-0 text-center font-sans text-4xl font-semibold tracking-wider text-accent1">
        {operatorMap[props.data.func]}
      </h2>
    </DefaultNode>
  );
};

export default memo(ArithmeticNode);
