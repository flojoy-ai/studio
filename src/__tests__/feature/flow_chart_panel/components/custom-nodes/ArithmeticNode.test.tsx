import { renderWithTheme } from "@src/__tests__/__utils__/utils";
import ArithmeticNode from "@src/feature/flow_chart_panel/components/custom-nodes/ArithmeticNode";
import { CustomNodeProps } from "@src/feature/flow_chart_panel/types/CustomNodeProps";

const props: CustomNodeProps = {
  data: { id: "test-id", label: "test", func: "test", type: "test", ctrls: {} },
};

jest.mock("@hooks/useFlowChartState");

jest.mock("@feature/flow_chart_panel/components/HandleComponent", () => {
  const mockChildren = jest
    .fn()
    .mockReturnValue(<div data-testid="handle-component" />);
  return { __esModule: true, default: mockChildren };
});

describe("ArithmeticNode", () => {
  it("checks the snapshot", () => {
    const { container } = renderWithTheme(<ArithmeticNode {...props} />);
    expect(container).toMatchSnapshot();
  });
  it.each([
    ["MULTIPLY", "multiply-svg"],
    ["ADD", "add-svg"],
    ["SUBTRACT", "sub-svg"],
    ["default", "default-svg"],
    ["handle", "handle-component"],
  ])("checks if component: %p in the document", (funcName, testId) => {
    let component;

    if (funcName !== "default") {
      const { getByTestId } = renderWithTheme(
        <ArithmeticNode data={{ ...props.data, func: funcName }} />
      );
      component = getByTestId(testId);
    } else {
      const { getByTestId } = renderWithTheme(<ArithmeticNode {...props} />);
      component = getByTestId(testId);
    }

    expect(component).toBeInTheDocument();
  });
});
