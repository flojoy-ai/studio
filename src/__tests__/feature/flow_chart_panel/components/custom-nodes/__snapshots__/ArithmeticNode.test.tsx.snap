// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArithmeticNode checks the snapshot 1`] = `
<div>
  <div
    class="mantine-n8y7xk"
    data-testid="node-wrapper"
  >
    <div
      class="mantine-1avyp1d"
    >
      <div
        class="mantine-18dvbbq"
      >
        <svg
          data-testid="default-svg"
          fill="none"
          height="100"
          viewBox="0 0 72 81"
          width="99"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M69.7184 43.874L6.41934 79.9997C4.08605 81.3313 1.18449 79.6464 1.18449 76.9599V4.70849C1.18449 2.02195 4.08604 0.337065 6.41933 1.66871L69.7184 37.7944C72.072 39.1376 72.072 42.5308 69.7184 43.874Z"
            fill="#FFC93F"
            fill-opacity="0.2"
            stroke="#FFC93F"
          />
        </svg>
        <div
          class="mantine-1avyp1d"
        />
        <div
          class="mantine-x06mmk"
        >
          <div
            data-testid="handle-component"
          />
        </div>
      </div>
    </div>
  </div>
  ,
</div>
`;
